model/task_lifecycle.go

```diff
@@ -1643,18 +1643,27 @@ func ClearAndResetStrandedHostTask(h *host.Host) error {
 	return nil
 }
 
-// ResetStaleTask fixes a task that has exceeded the heartbeat timeout by
-// marking the current task execution as finished and, if possible, a new
-// execution is created to restart the task.
+// ResetStaleTask fixes a task that has either exceeded the heartbeat timeout
+// or has been marked as aborted but was never ended by the agent.
+// The current task execution is marked as finished and, if the task is not
+// aborted, a new execution is created to restart the task.
 func ResetStaleTask(t *task.Task) error {
 	CheckAndBlockSingleHostTaskGroup(t, t.Status)
 
-	if err := resetSystemFailedTask(t, evergreen.TaskDescriptionHeartbeat); err != nil {
-		return errors.Wrap(err, "resetting task")
+	failureDesc := evergreen.TaskDescriptionHeartbeat
+	if t.Aborted {
+		failureDesc = evergreen.TaskDescriptionAborted
+		if err := t.MarkSystemFailed(failureDesc); err != nil {
+			return errors.Wrap(err, "marking aborted task as system failed")
+		}
```

Comment:
I'm slightly confused how this will work if we're assuming that the agent end task route must be called for this abort to work. My question here is, if the task is not reset, we're requiring that the end task logic for this task will happen sometime. is it guaranteed that all the end task logic will actually happen? If the stale heartbeat job was triggered, it seems like the host/agent might already be in a bad state, so it might never hit the end task route to do the end task updates.

Reply:
This is meant to capture the case where a task is marked as `Aborted` in the task doc, but no agent ever hits the end task route to do the end task updates within the defined heartbeat timeout threshold. 

It's meant to perform cleanup on the bad state of these tasks and the logic is meant to be analagous to how the heartbeat tasks are cleaned up on L1660.  The only difference is that after the heartbeat tasks are marked system failed, they then get automatically reset and create a new execution, whereas if a task is marked as aborted we should just mark it as ended due to abort and stop there.

Reply:
I guess my question was, since `(*Task).MarkSystemFailed` just updates this one task and will not reset it to try again, should we expect this to do all the other task-ending logic as well (like update the commit queue, update dependencies, reset when finished, etc.)?

Reply:
Gotcha, I see.  Ok in this case the work you described that the agent would normally do that is not happening here is in `MarkEnd`.  

The execution timeout job already performs `CheckAndBlockSingleHostTaskGroup` in the same way the agent's end task route does, and `HandleEndTaskForCommitQueueTask` is a no-op for failed and aborted tasks so I've omitted that. The only change that I've made is to call `MarkEnd` rather than `(*Task).MarkSystemFailed`, which handles all of the dependency updating, reset when finished, etc logic.